runtime/common/src/check_nonce.rs

```diff
@@ -0,0 +1,347 @@
+			is_eth_tx: false,
+			eth_tx_valid_until: Zero::zero(),
+		}
+	}
+
+	pub fn mark_as_ethereum_tx(&mut self, valid_until: T::BlockNumber) {
+		self.is_eth_tx = true;
+		self.eth_tx_valid_until = valid_until;
+	}
+}
+
+impl<T: frame_system::Config + module_evm::Config> sp_std::fmt::Debug for CheckNonce<T> {
+	#[cfg(feature = "std")]
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		write!(
+			f,
+			"CheckNonce(nonce: {}, is_eth_tx: {} eth_tx_valid_until: {})",
+			self.nonce, self.is_eth_tx, self.eth_tx_valid_until
+		)
+	}
+
+	#[cfg(not(feature = "std"))]
+	fn fmt(&self, _: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+		Ok(())
+	}
+}
+
+impl<T: frame_system::Config + module_evm::Config> SignedExtension for CheckNonce<T>
+where
+	T::Call: Dispatchable<Info = DispatchInfo>,
+	T::AddressMapping: AddressMapping<T::AccountId>,
+{
+	type AccountId = T::AccountId;
+	type Call = T::Call;
+	type AdditionalSigned = ();
+	type Pre = ();
+	const IDENTIFIER: &'static str = "CheckNonce";
+
+	fn additional_signed(&self) -> sp_std::result::Result<(), TransactionValidityError> {
+		Ok(())
+	}
+
+	fn pre_dispatch(
+		self,
+		who: &Self::AccountId,
+		_call: &Self::Call,
+		_info: &DispatchInfoOf<Self::Call>,
+		_len: usize,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		let mut account = frame_system::Account::<T>::get(who);
+		if self.is_eth_tx {
+			// should check evm nonce
+			let address = <T as module_evm::Config>::AddressMapping::get_evm_address(who)
+				.unwrap_or_else(|| <T as module_evm::Config>::AddressMapping::get_default_evm_address(who));
+			let evm_nonce = module_evm::Accounts::<T>::get(&address)
+				.map(|x| x.nonce)
+				.unwrap_or_default();
+			if self.nonce != evm_nonce {
+				return Err(if self.nonce < evm_nonce {
+					InvalidTransaction::Stale
+				} else {
+					InvalidTransaction::Future
+				}
+				.into());
+			}
+		} else {
+			if self.nonce != account.nonce {
+				return Err(if self.nonce < account.nonce {
+					InvalidTransaction::Stale
+				} else {
+					InvalidTransaction::Future
+				}
+				.into());
+			}
+		}
+		account.nonce += T::Index::one();
+		frame_system::Account::<T>::insert(who, account);
+		Ok(())
+	}
+
+	fn validate(
+		&self,
+		who: &Self::AccountId,
+		_call: &Self::Call,
+		_info: &DispatchInfoOf<Self::Call>,
+		_len: usize,
+	) -> TransactionValidity {
+		if self.is_eth_tx {
+			// should check evm nonce
+			let address = <T as module_evm::Config>::AddressMapping::get_evm_address(who)
+				.unwrap_or_else(|| <T as module_evm::Config>::AddressMapping::get_default_evm_address(who));
+			let evm_nonce = module_evm::Accounts::<T>::get(&address)
+				.map(|x| x.nonce)
+				.unwrap_or_default();
+			if self.nonce < evm_nonce {
+				return InvalidTransaction::Stale.into();
+			}
+
+			let provides = vec![Encode::encode(&(who, self.nonce, b"eth_tx"))];
+			let requires = if evm_nonce < self.nonce {
+				vec![Encode::encode(&(who, self.nonce - One::one(), b"eth_tx"))]
```

Comment:
```suggestion
			let provides = vec![Encode::encode(&(address, self.nonce, b"eth_tx"))];
			let requires = if evm_nonce < self.nonce {
				vec![Encode::encode(&(address, self.nonce - One::one(), b"eth_tx"))]
```

@xlc Should I need to use `address` instead of `who`?

Reply:
this code is only triggered with eth signature so best to use the eth address

Reply:
we might not need `b"eth_tx"` if eth address is used?